src/include/common/worker_pool.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+#include <atomic>
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <iostream>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <string>
+#include <thread>  // NOLINT
+#include <utility>
+#include <vector>
+
+namespace terrier::common {
+
+using TaskQueue = std::queue<std::function<void()>>;
+
+/**
+ * A worker pool that maintains a group of worker threads and a task queue.
+ *
+ * As soon as there is a task in the task queue, a worker thread will be
+ * assigned to run that task. A task must be a function that takes no argument.
+ * After a worker finishes a task, it will eagerly try to get a new task.
+ *
+ * This pool is restartable, meaning it can be started again after it has been
+ * shutdown. Calls to Startup() and Shutdown() are thread-safe.
+ */
+class WorkerPool {
+ public:
+  /**
+   * Initialize the worker pool. Once the number of worker is set by the constructor
+   * it cannot be changed.
+   *
+   * After initialization of the worker pool with a given task queue, worker threads
+   * do NOT start running. Need call StartUp() to start working on tasks.
+   *
+   * @param pool_name the name of the worker pool
+   * @param num_workers the number of workers in this pool
+   * @param task_queue a queue of tasks
+   */
+  WorkerPool(std::string pool_name, uint32_t num_workers, TaskQueue task_queue)
+      : pool_name_(std::move(pool_name)),
+        num_workers_(num_workers),
+        is_running_(false),
+        task_queue_(std::move(task_queue)) {}
+
+  /**
+   * Destructor. Wake up all workers and let them finish before it's destroyed.
+   */
+  ~WorkerPool() {
+    std::unique_lock<std::mutex> lock(task_lock_);  // grab the lock
+    is_running_ = false;                            // signal all the threads to shutdown
+    task_cv_.notify_all();                          // wake up all the threads
+    lock.unlock();                                  // free the lock
+    for (auto &thread : workers_) thread.join();
+  }
+
+  /**
+   * Start the worker pool. If the task queue is empty or we run out of tasks,
+   * workers will be put into sleep.
+   */
+  void Startup() {
```

Comment:
Is there a use case where we want a thread pool that is not started? If not, maybe it is better for the constructor to encapsulate this behavior.

Reply:
A thread pool can be shut down. So we would like a way to start it up again.

Reply:
You can keep this method. I was more asking whether it makes sense to always just call it in the constructor body?